src/packet.rs

```diff
@@ -0,0 +1,80 @@
+//! Implements serializing and deserializing MIDI messages as universal midi packets (UMP)
+use core::ops::Deref;
+
+/// A universal midi packet (UMP) is a 32, 64, 96, or 128 bit slice of serialized
+/// MIDI data that is parsed into midi messages, or serialized from them.
+///
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+pub struct Packet<const N: usize>(pub(crate) [u32; N]);
+
+impl<const N: usize> Deref for Packet<N> {
+    type Target = [u32];
+    fn deref(&self) -> &'_ Self::Target {
+        &self.0
+    }
+}
+
+impl<const N: usize> Packet<N> {
+    pub fn message_type(&self) -> u8 {
+        self.0[0].to_ne_bytes()[0] >> 4
+    }
+
+    pub fn group(&self) -> u8 {
+        self.0[0].to_ne_bytes()[0] & 0x0f
+    }
+
+    pub fn status(&self) -> u8 {
+        // Lead
+        self.0[0].to_ne_bytes()[1]
+    }
+}
+
+pub type Packet32 = Packet<1>;
+pub type Packet64 = Packet<2>;
+pub type Packet96 = Packet<3>;
+pub type Packet128 = Packet<4>;
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+#[repr(u8)]
+pub enum MessageType {
+    Utility = 0,
+    System = 1,
+    LegacyChannelVoice = 2,
+    Data64 = 3,
+    ChannelVoice = 4,
+    Data128 = 5,
+    Reserved6 = 6,
+    Reserved7 = 7,
+    Reserved8 = 8,
+    Reserved9 = 9,
+    Reserved10 = 10,
+    Reserved11 = 11,
+    Reserved12 = 12,
+    Reserved13 = 13,
+    Reserved14 = 14,
+    Reserved15 = 15,
+}
+
+impl From<u8> for MessageType {
+    fn from(value: u8) -> Self {
+        match value {
+            0 => MessageType::Utility,
+            1 => MessageType::System,
+            2 => MessageType::LegacyChannelVoice,
+            3 => MessageType::Data64,
+            4 => MessageType::ChannelVoice,
+            5 => MessageType::Data128,
+            6 => MessageType::Reserved6,
+            7 => MessageType::Reserved7,
+            8 => MessageType::Reserved8,
+            9 => MessageType::Reserved9,
+            10 => MessageType::Reserved10,
+            11 => MessageType::Reserved11,
+            12 => MessageType::Reserved12,
+            13 => MessageType::Reserved13,
+            14 => MessageType::Reserved14,
+            15 => MessageType::Reserved15,
+            _ => unreachable!("Invalid value for message type."),
```

Comment:
Hmm. This makes me wonder.
This is really a type error, not a value error (the type is u4 that happens to be represented as u8).
It's not the first time I come to be niggled by this situation – each place where I consider conversion from byte to a type that is really represented as a nibble, I find it to be ambiguous. It's not clear whether I should be using the whole byte, or if I have to prepare it (bit shift it) beforehand. 
Nibbles are used very often in the specification and I'm sure this will continue when the reserved message types are defined. It might be a pattern that's worth representing in the type system. Maybe have a `struct U4(u8)` to enforce this assumption by the type system? It would make parsing of the more convoluted messages simpler, I think.
Or have you already considered this?